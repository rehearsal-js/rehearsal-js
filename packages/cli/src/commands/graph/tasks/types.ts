export interface GraphWorkerOptions {
  rootPath: string;
  srcPath: string;
  ignore: string[];
  serviceMap: Record<string, string>;
  output?: string;
}

type MessageResponse = { type: 'message'; content: string };
type FilesResponse = { type: 'files'; content: string[] };
export type Response = MessageResponse | FilesResponse;
