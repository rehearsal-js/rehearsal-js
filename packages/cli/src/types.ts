import { MigrationStrategy } from '@rehearsal/migration-graph';

import { UserConfig } from './userConfig';
import { State } from './helpers/state';

export type CliCommand = 'upgrade' | 'migrate';
export type Formats = 'sarif' | 'json' | 'sonarqube' | 'md';

export type MigrateCommandOptions = {
  basePath: string;
  entrypoint: string;
  files: string;
  format: Formats[];
  outputPath: string;
  verbose: boolean | undefined;
  userConfig: string | undefined;
  interactive: boolean | undefined;
};

export type MigrateCommandContext = {
  skip: boolean;
  userConfig: UserConfig | undefined;
  strategy: MigrationStrategy;
  sourceFilesWithAbsolutePath: string[];
  sourceFilesWithRelativePath: string[];
  input: unknown;
  targetPackagePath: string;
  state: State;
};

export type UpgradeCommandContext = {
  tsVersion: string;
  latestAvailableBuild: string;
  currentTSVersion: string;
  skip: boolean;
};

export type UpgradeCommandOptions = {
  build: string;
  basePath: string;
  format: Formats[];
  outputPath: string;
  tsVersion: string;
  dryRun: boolean;
};

export type DependencyConfig = {
  dependencies?: string[];
  devDependencies?: string[];
};

export type SetupConfigCommand = {
  command: string;
  args: string[];
};

export type SetupConfig = {
  ts?: SetupConfigCommand;
  lint?: SetupConfigCommand;
};

export type CustomCommandConfig = {
  install?: DependencyConfig;
  setup?: SetupConfig;
};

export type CustomConfig = {
  upgrade?: CustomCommandConfig;
  migrate?: CustomCommandConfig;
};

export type PackageSelection = {
  name: string;
  path: string;
};

export type MenuMap = {
  [key: string]: string;
};

export type MigrationSummary = {
  totalErrorCount: number;
  hintAddedCount: number;
};
export type TSConfig = {
  compilerOptions: {
    strict: boolean;
  };
};

export type ScriptMap = {
  [key: string]: string;
};
