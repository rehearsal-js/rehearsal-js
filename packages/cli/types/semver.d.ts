declare module 'semver' {
  const semver: {
    valid: (input: string) => boolean;
  };
  export = semver;
}
