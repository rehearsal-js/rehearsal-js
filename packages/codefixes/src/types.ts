import {
  CodeFixAction,
  DiagnosticWithLocation,
  FileTextChanges,
  LanguageService,
  Node,
  Program,
  TypeChecker,
} from 'typescript';

export interface DiagnosticWithContext extends DiagnosticWithLocation {
  service: LanguageService;
  program: Program;
  checker: TypeChecker;
  node?: Node;
}

export interface CodeHintList {
  [key: number]: CodeHintListElement;
}

export interface CodeHintListElement {
  hint: string;
  helpUrl?: string;
  hints?: CodeHint[];
}

export interface CodeHint {
  when: (n: Node, p: Program, c: TypeChecker) => boolean;
  hint: string;
}

export interface CodeFixCollection {
  getFixForDiagnostic(diagnostic: DiagnosticWithContext): CodeFixAction | undefined;
}

export type CodeFixKind = 'insert' | 'delete' | 'replace';

export interface FixedFile {
  fileName: string;
  newCode?: string;
  oldCode?: string;
  codeFixAction?: CodeFixKind;
  location: {
    startLine: number;
    startColumn: number;
    endLine: number;
    endColumn: number;
  };
}

export interface CodeFix {
  getCodeAction: (diagnostic: DiagnosticWithContext) => CodeFixAction | undefined;
}

export function createCodeFixAction(
  fixName: string,
  changes: FileTextChanges[],
  description: string
): CodeFixAction {
  return {
    fixName,
    description,
    changes,
  };
}
