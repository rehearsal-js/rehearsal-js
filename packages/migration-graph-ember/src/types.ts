/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Package } from '@rehearsal/migration-graph-shared';
import type { EmberAddonPackage } from './entities/ember-addon-package.js';
import type { EmberAppPackage } from './entities/ember-app-package.js';

export type EmberProjectPackage = Package | EmberAddonPackage | EmberAppPackage;
