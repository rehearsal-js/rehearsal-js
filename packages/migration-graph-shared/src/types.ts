import type { Package } from './entities/package.js';

export type UniqueNode = {
  key: string;
  synthetic?: boolean;
};

export type PackageNode = UniqueNode & {
  pkg: Package;
  converted?: boolean;
};

export type ModuleNode = UniqueNode & {
  path: string;
  parsed?: boolean;
};
