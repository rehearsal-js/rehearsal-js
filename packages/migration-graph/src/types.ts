import { Package } from '@rehearsal/migration-graph-shared';

import { Graph } from './utils/graph';

export type UniqueGraphNode = {
  key: string;
};

export type PackageNode = UniqueGraphNode & {
  pkg: Package;
  converted?: boolean;
  modules: Graph<ModuleNode>;
};

export type ModuleNode = UniqueGraphNode & {
  path: string;
  parsed?: boolean;
  meta?: unknown;
};
