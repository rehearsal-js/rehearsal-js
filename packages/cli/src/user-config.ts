import { resolve } from 'node:path';
import { addDep } from '@rehearsal/utils';
import { execa } from 'execa';
import { readJSONSync } from 'fs-extra/esm';
import type {
  MigrateCommandConfig,
  UpgradeCommandConfig,
  CliCommand,
  CustomConfig,
} from './types.js';

// Storage and runner for user custom cli config
export class UserConfig {
  public basePath: string;
  public config?: MigrateCommandConfig | UpgradeCommandConfig;

  constructor(basePath: string, command: CliCommand) {
    const config: CustomConfig = readJSONSync(resolve(basePath, 'rehearsal-config.json'));

    this.config = config[command];

    this.basePath = basePath;
  }

  public get hasDependencies(): boolean {
    return (
      !!this.config?.install &&
      (!!this.config?.install?.dependencies || !!this.config?.install?.devDependencies)
    );
  }

  public get dependencies(): string[] {
    if (this.config && this.config.install) {
      const { dependencies } = this.config.install;
      if (dependencies && dependencies.length) {
        return dependencies;
      }
    }
    return [];
  }

  public get devDependencies(): string[] {
    if (this.config && this.config.install) {
      const { devDependencies } = this.config.install;
      if (devDependencies && devDependencies.length) {
        return devDependencies;
      }
    }
    return [];
  }

  public get hasPostInstallHook(): boolean {
    return !!this.config?.postInstall;
  }

  public get hasTsSetup(): boolean {
    return !!this.config?.setup?.ts;
  }

  public get hasPostTsSetupHook(): boolean {
    return !!this.config?.setup?.postTsSetup;
  }

  public get hasLintSetup(): boolean {
    return !!this.config?.setup?.lint;
  }

  public get hasPostLintSetup(): boolean {
    return !!this.config?.setup?.postLintSetup;
  }

  async install(): Promise<void> {
    if (this.config && this.config.install) {
      const { dependencies, devDependencies } = this.config.install;
      if (dependencies && dependencies.length) {
        await addDep(dependencies, false, { cwd: this.basePath });
      }

      if (devDependencies && devDependencies.length) {
        await addDep(devDependencies, true, { cwd: this.basePath });
      }
    }
  }

  async postInstall(): Promise<void> {
    if (this.config && this.config.postInstall) {
      if (Array.isArray(this.config.postInstall)) {
        for (const { command, args } of this.config.postInstall) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = this.config.postInstall;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  // TODO: how to make setup tasks more generic instead of separated functions?

  // TODO: it doesn't make sense for migrate and upgrade sharing the same setup tasks,
  // What does upgrade command need for custom config?
  async tsSetup(): Promise<void> {
    if (this.config?.setup?.ts) {
      const { ts: tsSetup } = this.config.setup;
      if (Array.isArray(tsSetup)) {
        for (const { command, args } of tsSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = tsSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async postTsSetup(): Promise<void> {
    if (this.config?.setup?.postTsSetup) {
      const { postTsSetup } = this.config.setup;
      if (Array.isArray(postTsSetup)) {
        for (const { command, args } of postTsSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = postTsSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async lintSetup(): Promise<void> {
    if (this.config?.setup?.lint) {
      const { lint: lintSetup } = this.config.setup;
      if (Array.isArray(lintSetup)) {
        for (const { command, args } of lintSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = lintSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  async postLintSetup(): Promise<void> {
    if (this.config?.setup?.postLintSetup) {
      const { postLintSetup } = this.config.setup;
      if (Array.isArray(postLintSetup)) {
        for (const { command, args } of postLintSetup) {
          await execa(command, args, { cwd: this.basePath });
        }
      } else {
        const { command, args } = postLintSetup;
        await execa(command, args, { cwd: this.basePath });
      }
    }
  }

  get exclude(): string[] | undefined {
    return (this.config as MigrateCommandConfig)?.exclude;
  }

  get include(): string[] | undefined {
    return (this.config as MigrateCommandConfig)?.include;
  }
}
