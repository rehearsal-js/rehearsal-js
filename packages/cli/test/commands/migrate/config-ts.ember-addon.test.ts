import { resolve } from 'node:path';
import { readdirSync } from 'node:fs';
import { afterEach, beforeEach, afterAll, describe, expect, test, vi } from 'vitest';
import { readJSONSync, writeJSONSync } from 'fs-extra/esm';
import { getEmberAddonProject } from '@rehearsal/test-support';
import { Project } from 'fixturify-project';
import { tsConfigTask } from '../../../src/commands/migrate/tasks/index.js';
import { listrTaskRunner, createMigrateOptions } from '../../test-helpers/index.js';
import { runTsConfig, createUserConfig } from '../../test-helpers/config-ts-test-utils.js';
import { TSConfig } from '../../../src/types.js';
import { UserConfig } from '../../../src/user-config.js';

const projects = {
  emberAddon: getEmberAddonProject(),
};

describe('Task: config-ts ember app', () => {
  for (const [name, originalProject] of Object.entries(projects)) {
    describe(name, () => {
      let project: Project;
      let output = '';

      beforeEach(async () => {
        project = originalProject.clone();
        await project.write();

        output = '';
        vi.spyOn(console, 'info').mockImplementation((chunk: string) => {
          chunk = chunk.replace(new RegExp(`${project.baseDir}`, 'g'), '<tmp-path>');
          output += `${chunk}\n`;
        });

        vi.spyOn(console, 'log').mockImplementation((chunk: string) => {
          chunk = chunk.replace(new RegExp(`${project.baseDir}`, 'g'), '<tmp-path>');
          output += `${chunk}\n`;
        });
      });

      afterEach(() => {
        vi.clearAllMocks();
        output = '';
        project.dispose();
      });

      afterAll(() => {
        originalProject.dispose();
      });

      test('create tsconfig if not existed', async () => {
        await runTsConfig(project.baseDir);

        expect(readJSONSync(resolve(project.baseDir, 'tsconfig.json'))).matchSnapshot();
        expect(output).matchSnapshot();
      });

      test('update tsconfig if exists', async () => {
        // Prepare old tsconfig
        const oldTsConfig = { compilerOptions: { strict: false } };
        writeJSONSync(resolve(project.baseDir, 'tsconfig.json'), oldTsConfig);

        await runTsConfig(project.baseDir);

        const tsConfig = readJSONSync(resolve(project.baseDir, 'tsconfig.json')) as TSConfig;

        expect(tsConfig.compilerOptions.strict).toBeTruthy();
        expect(output).toMatchSnapshot();
      });

      test('update tsconfig if invalid extends exist', async () => {
        // Prepare old tsconfig
        const oldTsConfig = { extends: 'invalid-tsconfig.json' };
        writeJSONSync(resolve(project.baseDir, 'tsconfig.json'), oldTsConfig);

        await runTsConfig(project.baseDir);

        const tsConfig = readJSONSync(resolve(project.baseDir, 'tsconfig.json')) as TSConfig;

        expect(tsConfig.compilerOptions.strict).toBeTruthy();
        expect(output).toMatchSnapshot();
      });

      test('skip if tsconfig.json exists with strict on', async () => {
        // Prepare old tsconfig
        const oldTsConfig = { compilerOptions: { strict: true } };
        writeJSONSync(resolve(project.baseDir, 'tsconfig.json'), oldTsConfig);

        await runTsConfig(project.baseDir);
        await runTsConfig(project.baseDir); //should be skipped

        expect(output).toMatchSnapshot();
      });

      test('run custom config command with user config provided', async () => {
        createUserConfig(project.baseDir, {
          migrate: {
            setup: {
              ts: { command: 'touch', args: ['custom-ts-config-script'] },
            },
          },
        });

        const options = createMigrateOptions(project.baseDir, {
          userConfig: 'rehearsal-config.json',
        });
        const userConfig = new UserConfig(project.baseDir, 'rehearsal-config.json', 'migrate');
        const tasks = [tsConfigTask(options, { userConfig })];
        await listrTaskRunner(tasks);

        // This proves the custom command works
        expect(readdirSync(project.baseDir)).toContain('custom-ts-config-script');
        expect(output).toMatchSnapshot();
      });

      test('skip custom config command', async () => {
        // Prepare old tsconfig
        const oldTsConfig = { compilerOptions: { strict: true } };
        writeJSONSync(resolve(project.baseDir, 'tsconfig.json'), oldTsConfig);

        createUserConfig(project.baseDir, {
          migrate: {
            setup: {
              ts: { command: 'touch', args: ['custom-ts-config-script'] },
            },
          },
        });

        await runTsConfig(project.baseDir, { userConfig: 'rehearsal-config.json' });

        // This proves the custom command works not triggered
        expect(readdirSync(project.baseDir)).not.toContain('custom-ts-config-script');
        expect(output).toMatchSnapshot();
      });

      test('postTsSetup hook from user config', async () => {
        createUserConfig(project.baseDir, {
          migrate: {
            setup: {
              ts: { command: 'touch', args: ['custom-ts-config-script'] },
              postTsSetup: { command: 'mv', args: ['custom-ts-config-script', 'foo'] },
            },
          },
        });

        const options = createMigrateOptions(project.baseDir, {
          userConfig: 'rehearsal-config.json',
        });
        const userConfig = new UserConfig(project.baseDir, 'rehearsal-config.json', 'migrate');
        const tasks = [tsConfigTask(options, { userConfig })];
        await listrTaskRunner(tasks);

        expect(readdirSync(project.baseDir)).toContain('foo');
        expect(readdirSync(project.baseDir)).not.toContain('custom-ts-config-script');
        expect(output).toMatchSnapshot();
      });
    });
  }
});
