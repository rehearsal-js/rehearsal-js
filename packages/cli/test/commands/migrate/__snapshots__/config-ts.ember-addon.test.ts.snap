// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task: config-ts ember app > emberAddon > create tsconfig if not existed 1`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "declaration": false,
    "experimentalDecorators": true,
    "importsNotUsedAsValues": "error",
    "inlineSourceMap": true,
    "inlineSources": true,
    "isolatedModules": true,
    "noEmit": true,
    "noImplicitOverride": true,
    "noUncheckedIndexedAccess": true,
    "preserveValueImports": true,
    "strict": true,
    "target": "es2020",
  },
  "display": "Ember App",
  "extends": "../base.tsconfig.json",
  "glint": {
    "environment": {
      "ember-loose": {},
      "ember-template-imports": {},
    },
  },
}
`;

exports[`Task: config-ts ember app > emberAddon > create tsconfig if not existed 2`] = `
"[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
"
`;

exports[`Task: config-ts ember app > emberAddon > do not skip if custom config exists 1`] = `
"[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
"
`;

exports[`Task: config-ts ember app > emberAddon > postTsSetup hook from user config 1`] = `
"[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[DATA] Run postTsSetup from config
[SUCCESS] Create tsconfig.json
"
`;

exports[`Task: config-ts ember app > emberAddon > run custom config command with user config provided 1`] = `
"[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
"
`;

exports[`Task: config-ts ember app > emberAddon > skip if tsconfig.json exists with strict on 1`] = `
"[STARTED] Create tsconfig.json
[SKIPPED] Create tsconfig.json
[STARTED] Create tsconfig.json
[SKIPPED] Create tsconfig.json
"
`;

exports[`Task: config-ts ember app > emberAddon > update tsconfig if exists 1`] = `
"[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
"
`;

exports[`Task: config-ts ember app > emberAddon > update tsconfig if invalid extends exist 1`] = `
"[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
"
`;
